core/src/main/scala/kafka/log/TimeIndex.scala

```diff
@@ -0,0 +1,198 @@
+          .format(offset, _entries, _lastOffset, file.getAbsolutePath))
+      // We only append to the time index when the timestamp is greater than the last inserted timestamp.
+      // If the messages are in message format v0, the timestamp will always be NoTimestamp. In that case, the time
+      // index will be empty until the segment is closed. When closing the segment, the broker will insert a time
+      // index entry whose timestamp is the last modification time of the file, and the entry points to the base offset.
+      if (timestamp > _lastTimestamp) {
+        debug("Adding index entry %d => %d to %s.".format(timestamp, offset, file.getName))
+        mmap.putLong(timestamp)
+        mmap.putInt((offset - baseOffset).toInt)
+        _entries += 1
+        _lastTimestamp = timestamp
+        _lastOffset = offset
+        require(_entries * entrySize == mmap.position, _entries + " entries but file position in index is " + mmap.position + ".")
+      }
+    }
+  }
+
+  /**
+   * Find the time index entries whose timestamp is less than the given timestamp and return a number of closest
+   * entries.
+   *
+   * @param targetTimestamp The timestamp to look up.
+   * @param numOffsets The maximum number of time index entries to return.
+   *
+   * @return A sequence of time index entries.
+   */
+  def lookup(targetTimestamp: Long, numOffsets: Int = 1): Seq[TimestampOffset] = {
+    maybeLock(lock) {
+      _entries match {
+        case 0 => Seq.empty
+        case _ =>
+          val idx = mmap.duplicate
+          val foundSlot = indexSlotFor(idx, targetKey = Some(targetTimestamp))
+          // The binary search will return the exact match when there is one. If the timestamp in the index entry is
+          // the same as target timestamp, we need to use the earlier entry.
+          val slot = {
+            if (foundSlot >= 0 && timestamp(idx, foundSlot) == targetTimestamp)
+              foundSlot - 1
+            else
+              foundSlot
+          }
+          if (slot == -1)
+            Seq.empty
+          else {
+            val entries = new ListBuffer[TimestampOffset]
+            for (i <- slot to math.max(0, slot - numOffsets + 1) by -1)
+              entries += parseEntry(idx, i).asInstanceOf[TimestampOffset]
+            entries.toSeq
+          }
+      }
+    }
+  }
+
+  override def truncate() = truncateToEntries(0)
+
+  /**
+   * Remove all entries from the index which have an offset greater than or equal to the given offset.
+   * Truncating to an offset larger than the largest in the index has no effect.
+   */
+  override def truncateTo(offset: Long) {
+    inLock(lock) {
+      val idx = mmap.duplicate
+      val slot = indexSlotFor(idx, targetValue = Some(offset))
+      /* There are 2 cases for choosing the new size
+       * 1) if there is no entry in the index <= the offset, delete everything
+       * 2) Otherwise delete everything larger than the closest smallest
```

Comment:
Hmm, not sure why this logic here is different from that in OffsetIndex.truncateTo. If we find an index entry that matches offset exactly, it seem that we should delete that slot from the time index too.